From da6852418d2d8bf9405219f888e103ccb8453f96 Mon Sep 17 00:00:00 2001
From: Yannick Cote <yannick.cote@qlogic.com>
Date: Fri, 19 Sep 2008 09:20:45 -0700
Subject: [PATCH] IB/ipath: ib_ipath module hangs on unload

If a send request is posted after the link is brought down, ib_ipath waits
on a CQE that never comes. This fix handles the case where posting a send
is requested when the link is down.

Signed-off-by: Yannick Cote <yannick.cote@qlogic.com>
---
 drivers/infiniband/hw/ipath/ipath_verbs.c |    7 +++++++
 1 files changed, 7 insertions(+), 0 deletions(-)

diff --git a/drivers/infiniband/hw/ipath/ipath_verbs.c b/drivers/infiniband/hw/ipath/ipath_verbs.c
index b766e40..eabc424 100644
--- a/drivers/infiniband/hw/ipath/ipath_verbs.c
+++ b/drivers/infiniband/hw/ipath/ipath_verbs.c
@@ -340,9 +340,16 @@ static int ipath_post_one_send(struct ipath_qp *qp, struct ib_send_wr *wr)
 	int acc;
 	int ret;
 	unsigned long flags;
+	struct ipath_devdata *dd = to_idev(qp->ibqp.device)->dd;
 
 	spin_lock_irqsave(&qp->s_lock, flags);
 
+	if (qp->ibqp.qp_type != IB_QPT_SMI &&
+	    !(dd->ipath_flags & IPATH_LINKACTIVE)) {
+		ret = -ENETDOWN;
+		goto bail;
+	}
+
 	/* Check that state is OK to post send. */
 	if (unlikely(!(ib_ipath_state_ops[qp->state] & IPATH_POST_SEND_OK)))
 		goto bail_inval;
-- 
1.6.0.2

